java-backend/src/main/java/org/kframework/backend/java/symbolic/SymbolicRewriter.java

```diff
@@ -691,72 +694,86 @@ private K disjunctResults(List<K> results) {
                     }
                 }*/
 
-                if (guarded) {
-                    ConstrainedTerm result = applySpecRules(term, specRules);
-                    if (result != null) {
+                    if (guarded) {
+                        ConstrainedTerm result = applySpecRules(term, specRules);
+                        if (result != null) {
+                            nextStepLogEnabled = true;
+                            logStep(step, v, term, true, alreadyLogged);
+                            // re-running constraint generation again for debug purposes
+                            if (global.javaExecutionOptions.logBasic) {
+                                System.err.println("\nApplying specification rule\n=========================\n");
+                            }
+                            if (visited.add(result)) {
+                                nextQueue.add(result);
+                            } else {
+                                if (term.equals(result)) {
+                                    throw KEMException.criticalError(
```

Comment:
I'm not sure this should be an error, as the specification has technically been proven true at this point.

Reply:
We already discussed this in some old PR. I don't see a valid use case for a spec rule that rewrites the term to itself.

Think of it like this: Any spec rule of the form `rule A => A` is technically valid, can be proved and represents no computation advancement. Without this check, once this rule matches during evaluation of another rule, that other rule gets proved as well. Why?

But it happened in the past that such rules were specification errors.

Reply:
I'm fine if it's just a matter of forbidding a spec rule that rewrites to itself, but my understanding (correct me if I'm wrong) is that this will also occur if the configuration rewrites something to itself, in which case proving the spec correct is a valid proof of that spec relative to that semantics. I would rather not make it impossible to prove properties about certain types of semantics in order to solve this problem. If you just want to forbid a spec that rewrites something to itself, that can be done as a preprocessing step in the k frontend without interfering with the ability to prove a nonterminating semantics.

Am I misunderstanding the scope here?

Reply:
It could happen in a less obvious fashion, like `rule <context> X => 0 </context>` where X matches 0. Can this be done at pre-processing phase? Is it worth the effort? But honestly this case was already discussed when I introduced the warning in the first place. Here I just converted the warning into error. @ehildenb was reviewer that time.

I don't understand your last comment. What you mean by "configuration rewrites something to itself". Can you give an example?

Reply:
So, what I mean is, suppose I have a semantics where a particular term doesn't terminate, and I want to prove that it doesn't terminate. Would not this change make such a proof impossible?

Reply:
Even so, you expect that return to a previously visited state will happen in more than 1 semantics step. By semantics steps I mean steps in which semantics rules are applied, not spec rules. If my change is in place, you would have to define spec rules in such a way that they don't capture one full loop, but a fragment of a loop. You'll have 2 rules, some `rule A=> B` and `rule B => A`, each of them encoding more than zero steps of the concrete semantics.

For the extremely unlikely case that we actually need to prove non-termination, this seems like a reasonable inconvenience. (I have arguments why it's unlikely, but don't want to get into this debate.) Much better then accidentally proving incorrect stuff due to easy to make errors in spec rules.